.github/workflows/ci.yml

```diff
@@ -23,33 +23,61 @@ jobs:
     strategy:
       fail-fast: false
       matrix:
-        os: [ubuntu-22.04, windows-2022]
         include:
-          - os: ubuntu-22.04
-            platform: linux
-          - os: windows-2022
-            platform: windows
+          - platform: linux
+            os: ubuntu-22.04
+            compiler: gcc-default
+          - platform: linux
+            os: ubuntu-22.04
+            compiler: gcc-latest
```

Comment:
C++ is supposed to be backwards compatible, so if things work with an older gcc version, they should work with the latest gcc version, right? In that sense, even though I understand this might come with some extra warnings and niceties, I'm not totally sure it is worth having another branch for it - at least not with every push.

I'm not sure how much you got in the standup yesterday about being environmental-aware when setting our CI systems (your connection was a bit unstable), so we don't waste resources and energy unnecessary.

Reply:
I didn't hear much so I was left wondering what you'd said :laughing:.

Yes, C++ is in theory backwards compatible. In practice though, you do often get compile errors when you move to a new compiler, usually because one header in your compiler's implementation of the standard library used to `#include` another and now it doesn't (e.g. #73). Compilers do often vary in how much of the various standards they implement and how faithfully too, which can also cause breakage (I just learned that the latest version of Apple's clang doesn't implement STL parallel algorithms *at all*, even though that's a C++17 feature, which means you can't compile Health-GPS with it). That's why I think it makes sense to have some minimum version of `gcc` that you test -- we're using Ubuntu LTS for the runner, so the default version included with that seems a sensible candidate -- but also to test against the latest version. That way we can be fairly confident that anyone using any of the intervening versions will be alright too. The extra warnings are a nice bonus, but also the Ubuntu folks tend to ship fairly old versions of packages, including `gcc`, so I do think it makes sense to test against the latest version of `gcc` too for people who use other distros with a less ancient version. (@jamesturner246 uses Fedora and I use Arch, which both ship newer versions than you find in Ubuntu LTS.) It also means that we don't have to fix a bunch of warnings and errors whenever we update the version of the Ubuntu runner. Admittedly, that's not a big job, but it would be better to fix the problems in advance, both for people using a newer version of `gcc` and also because compiler warnings can indicate things like serious memory corruption bugs caused by undefined behaviour -- unfortunately there are lots of ways you can shoot yourself in the foot with C++.

We are currently only running the CI for pushes to `main` and for PRs, so hopefully the resource usage isn't too terrible. The other thing we could do is restrict ourselves to building a release version of the code, not the debug version (see #122).

Reply:
What a nice explanation - worth a blogpost ;) . I hadn't realised that this runs only in pushes to main. Ok, no comments, then.